src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualChar.cs

```diff
@@ -3,64 +3,178 @@
+    /// VB. For example In C# someone may write: @"\z".  This should appear to the user the same as if they wrote "\\z"
+    /// and the same as "\\\u007a".  However, as these all have wildly different presentations for the user, there needs
+    /// to be a way to map back the characters it sees ( '\' and 'z' ) back to the  ranges of characters the user wrote.
+    /// </remarks>
+    internal readonly struct VirtualChar : IEquatable<VirtualChar>, IComparable<VirtualChar>, IComparable<char>
     {
-        public readonly char Char;
+        /// <summary>
+        /// The value of this <see cref="VirtualChar"/> as a <see cref="Rune"/> if such a represention is possible.
+        /// <see cref="Rune"/>s can represent Unicode codepoints that can appear in a <see cref="string"/> except for
+        /// unpaired surrogates.  If an unpaired high or low surrogate character is present, this value will be <see
+        /// cref="Rune.ReplacementChar"/>.  The value of this character can be retrieved from
+        /// <see cref="SurrogateChar"/>.
+        /// </summary>
+        public readonly Rune Rune;
+
+        /// <summary>
+        /// The unpaired high or low surrogate character that was encountered that could not be represented in <see
+        /// cref="Rune"/>.  If <see cref="Rune"/> is not <see cref="Rune.ReplacementChar"/> this will be <c>0</c>.
+        /// </summary>
+        public readonly char SurrogateChar;
+
+        /// <summary>
+        /// The span of characters in the original <see cref="SourceText"/> that represent this <see
+        /// cref="VirtualChar"/>.
+        /// </summary>
         public readonly TextSpan Span;
 
-        public VirtualChar(char @char, TextSpan span)
+        /// <summary>
+        /// Creates a new <see cref="VirtualChar"/> from the provided <paramref name="rune"/>.  This operation cannot
+        /// fail.
+        /// </summary>
+        public static VirtualChar Create(Rune rune, TextSpan span)
+            => new VirtualChar(rune, surrogateChar: default, span);
+
+        /// <summary>
+        /// Creates a new <see cref="VirtualChar"/> from an unpaired high or low surrogate character.  This will throw
+        /// if <paramref name="surrogateChar"/> is not actually a surrogate character. The resultant <see cref="Rune"/>
+        /// value will be <see cref="Rune.ReplacementChar"/>.
+        /// </summary>
+        public static VirtualChar Create(char surrogateChar, TextSpan span)
         {
+            if (!char.IsSurrogate(surrogateChar))
+                throw new ArgumentException(nameof(surrogateChar));
+
+            return new VirtualChar(rune: Rune.ReplacementChar, surrogateChar, span);
+        }
+
+        private VirtualChar(Rune rune, char surrogateChar, TextSpan span)
+        {
+            Contract.ThrowIfFalse(surrogateChar == 0 || rune == Rune.ReplacementChar,
+                "If surrogateChar is provided then rune must be Rune.ReplacementChar");
```

Comment:
💡 Consider making this an assertion to avoid the validation check in normal code paths. Since this is a private constructor, we should be able to guarantee this condition at all entry points.